In [2]:
import pandas as pd

df = pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]], columns=['A', 'B', 'C'],index = ['x','y','z'])
df.head()

A  B  C
x  1  2  3
y  4  5  6
z  7  8  9

In [3]:
df.tail(1)

A  B  C
z  7  8  9

In [4]:
df.index

Index(['x', 'y', 'z'], dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3 entries, x to z
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   A       3 non-null      int64
 1   B       3 non-null      int64
 2   C       3 non-null      int64
dtypes: int64(3)
memory usage: 96.0+ bytes


In [6]:
df.describe()

A    B    C
count  3.0  3.0  3.0
mean   4.0  5.0  6.0
std    3.0  3.0  3.0
min    1.0  2.0  3.0
25%    2.5  3.5  4.5
50%    4.0  5.0  6.0
75%    5.5  6.5  7.5
max    7.0  8.0  9.0

In [7]:
df.shape

(3, 3)

In [8]:
df.nunique() # number of unique values in each column

A    3
B    3
C    3
dtype: int64

In [9]:
#loading csv file
coffee = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/complete-pandas-tutorial/refs/heads/master/warmup-data/coffee.csv')
coffee.head()

Day Coffee Type  Units Sold
0     Monday    Espresso          25
1     Monday       Latte          15
2    Tuesday    Espresso          30
3    Tuesday       Latte          20
4  Wednesday    Espresso          35

We will often be using csv formats to read data. CSVs are great with readability.
However they are often heavy in terms of size - another alternative is parquet files which are often used in industry whn dealing with huge data. example a csv with 31.5 mb data can be compressed to parquet of 4.4 mb

In [10]:
results = pd.read_parquet('data/results.parquet')
results.head()

year    type discipline                     event                     as  \
0  1912.0  Summer     Tennis    Singles, Men (Olympic)  Jean-François Blanchy   
1  1912.0  Summer     Tennis    Doubles, Men (Olympic)  Jean-François Blanchy   
2  1920.0  Summer     Tennis    Singles, Men (Olympic)  Jean-François Blanchy   
3  1920.0  Summer     Tennis  Doubles, Mixed (Olympic)  Jean-François Blanchy   
4  1920.0  Summer     Tennis    Doubles, Men (Olympic)  Jean-François Blanchy   

   athlete_id  noc             team  place   tied medal  
0           1  FRA             None   17.0   True  None  
1           1  FRA   Jean Montariol    NaN  False  None  
2           1  FRA             None   32.0   True  None  
3           1  FRA  Jeanne Vaussard    8.0   True  None  
4           1  FRA  Jacques Brugnon    4.0  False  None

In [11]:
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          15
2     Tuesday    Espresso          30
3     Tuesday       Latte          20
4   Wednesday    Espresso          35
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [12]:
coffee.head(7)

Day Coffee Type  Units Sold
0     Monday    Espresso          25
1     Monday       Latte          15
2    Tuesday    Espresso          30
3    Tuesday       Latte          20
4  Wednesday    Espresso          35
5  Wednesday       Latte          25
6   Thursday    Espresso          40

In [13]:
coffee.tail(3)

Day Coffee Type  Units Sold
11  Saturday       Latte          35
12    Sunday    Espresso          45
13    Sunday       Latte          35

In [14]:
coffee.sample(5)

Day Coffee Type  Units Sold
13    Sunday       Latte          35
12    Sunday    Espresso          45
7   Thursday       Latte          30
0     Monday    Espresso          25
2    Tuesday    Espresso          30

In [15]:
#loc and iloc
coffee.loc[[0,1,8]]  #loc based on index 
coffee.loc[5:8]  #loc based on index range
coffee.loc[5:8,["Day", "Coffee Type"]] #loc based on index range and specific columns

Day Coffee Type
5  Wednesday       Latte
6   Thursday    Espresso
7   Thursday       Latte
8     Friday    Espresso

In [16]:
#iloc works preety much the same way but uses integer based indexing
coffee.iloc[[0,1,2]] #iloc based on integer index
coffee.iloc [1:4] #iloc based on range
coffee.iloc[1:4, 0:2] #iloc based on range for specific ronws and columns
#coffee.loc[1:4, 0:2] --- loc based on column range indexing --- throws error


Day Coffee Type
1   Monday       Latte
2  Tuesday    Espresso
3  Tuesday       Latte

In [17]:
#modifying using  loc and iloc
coffee.loc [2,"Units Sold"] = 56
coffee.loc [2:4, "Units Sold"] = 100

In [18]:
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          15
2     Tuesday    Espresso         100
3     Tuesday       Latte         100
4   Wednesday    Espresso         100
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [19]:
#we can also use at and iat for faster access of single elements. But this will not work for ranges
coffee.at[3,"Coffee Type"] = "Espresso"
coffee.iat[4,2] = 150

In [20]:
#For accessing columns
coffee["Coffee Type"]  #accessing a single column
coffee[["Coffee Type", "Units Sold"]]  #accessing multiple columns
coffee.Day  #accessing single column using dot notation -- works only if column name has no spaces

0        Monday
1        Monday
2       Tuesday
3       Tuesday
4     Wednesday
5     Wednesday
6      Thursday
7      Thursday
8        Friday
9        Friday
10     Saturday
11     Saturday
12       Sunday
13       Sunday
Name: Day, dtype: object

In [21]:
#sorting 
coffee.sort_values(by = "Units Sold", ascending = False)
coffee.sort_values(by =["Units Sold", "Day"], ascending = [True, False]) #sorting by multiple columns

Day Coffee Type  Units Sold
1      Monday       Latte          15
5   Wednesday       Latte          25
0      Monday    Espresso          25
7    Thursday       Latte          30
13     Sunday       Latte          35
11   Saturday       Latte          35
9      Friday       Latte          35
6    Thursday    Espresso          40
12     Sunday    Espresso          45
10   Saturday    Espresso          45
8      Friday    Espresso          45
2     Tuesday    Espresso         100
3     Tuesday    Espresso         100
4   Wednesday    Espresso         150

In [22]:
# iterating using for loop -- not very good in performance 
for index, rows in coffee.iterrows():
    print(index,rows["Day"], rows["Units Sold"])

# to iterate by columns using forloop 
for col in coffee:
    print(col)
    print(coffee[col])
    print

0 Monday 25
1 Monday 15
2 Tuesday 100
3 Tuesday 100
4 Wednesday 150
5 Wednesday 25
6 Thursday 40
7 Thursday 30
8 Friday 45
9 Friday 35
10 Saturday 45
11 Saturday 35
12 Sunday 45
13 Sunday 35
Day
0        Monday
1        Monday
2       Tuesday
3       Tuesday
4     Wednesday
5     Wednesday
6      Thursday
7      Thursday
8        Friday
9        Friday
10     Saturday
11     Saturday
12       Sunday
13       Sunday
Name: Day, dtype: object
Coffee Type
0     Espresso
1        Latte
2     Espresso
3     Espresso
4     Espresso
5        Latte
6     Espresso
7        Latte
8     Espresso
9        Latte
10    Espresso
11       Latte
12    Espresso
13       Latte
Name: Coffee Type, dtype: object
Units Sold
0      25
1      15
2     100
3     100
4     150
5      25
6      40
7      30
8      45
9      35
10     45
11     35
12     45
13     35
Name: Units Sold, dtype: int64


In [23]:
bios = pd.read_csv("Data/bios.csv")
bios.head()

athlete_id                   name   born_date    born_city  \
0           1  Jean-François Blanchy  1886-12-12     Bordeaux   
1           2         Arnaud Boetsch  1969-04-01       Meulan   
2           3           Jean Borotra  1898-08-13     Biarritz   
3           4        Jacques Brugnon  1895-05-11  Paris VIIIe   
4           5           Albert Canet  1878-04-17   Wandsworth   

            born_region born_country     NOC  height_cm  weight_kg   died_date  
0               Gironde          FRA  France        NaN        NaN  1960-10-02  
1              Yvelines          FRA  France      183.0       76.0         NaN  
2  Pyrénées-Atlantiques          FRA  France      183.0       76.0  1994-07-17  
3                 Paris          FRA  France      168.0       64.0  1978-03-20  
4               England          GBR  France        NaN        NaN  1930-07-25

In [24]:
#Filtering the data 
bios.loc[bios['height_cm']>200, ['name','height_cm']] 

name  height_cm
1080        Bozhil Lozanov      208.0
1673      Alexis Rubalcaba      204.0
2647         Milan Šrejber      203.0
2827    Nik'oloz Deriugini      206.0
2832         Elşad Qadaşev      205.0
...                    ...        ...
143784  Daniel Miklasevich      203.0
143833  Garrett Muagututia      205.0
143834         Kyle Ensing      201.0
143835         Mitch Stahl      203.0
143919   Emerson Rodríguez      204.0

[1219 rows x 2 columns]

In [25]:
#we canalso do something like this
bios[bios['height_cm'] >200][['name','height_cm']]

name  height_cm
1080        Bozhil Lozanov      208.0
1673      Alexis Rubalcaba      204.0
2647         Milan Šrejber      203.0
2827    Nik'oloz Deriugini      206.0
2832         Elşad Qadaşev      205.0
...                    ...        ...
143784  Daniel Miklasevich      203.0
143833  Garrett Muagututia      205.0
143834         Kyle Ensing      201.0
143835         Mitch Stahl      203.0
143919   Emerson Rodríguez      204.0

[1219 rows x 2 columns]

In [26]:
#multiple condition filtering
bios[(bios['height_cm'] > 200) & (bios['born_country'] == 'USA')]['name']

5197         Anne Donovan
5652          Mike Bantom
5657           Jim Barnes
5681         Walt Bellamy
5709           Larry Bird
               ...       
143706        Bam Adebayo
143707       Jerami Grant
143710    Khris Middleton
143712       JaVale McGee
143733         Nick Allen
Name: name, Length: 152, dtype: object

In [27]:
keith_bio = bios[bios['name'].str.contains("keith",case  = False)]
keith_bio.head()

athlete_id           name   born_date  born_city       born_region  \
1897         1907   Keith Hanlon  1966-09-01        NaN               NaN   
3505         3517  Keith Wallace  1961-03-29    Preston           England   
6228         6255  Keith Hartley  1940-10-15  Vancouver  British Columbia   
8898         8946    Keith Mwila  1966-01-01        NaN               NaN   
12053       12118   Keith Hervey  1898-11-03     Fulham           England   

      born_country            NOC  height_cm  weight_kg   died_date  
1897           NaN        Ireland        NaN        NaN         NaN  
3505           GBR  Great Britain      165.0       51.0  1999-12-31  
6228           CAN         Canada      200.0       85.0         NaN  
8898           NaN         Zambia        NaN        NaN  1993-01-09  
12053          GBR  Great Britain        NaN        NaN  1973-02-22

In [28]:
keith_first_names = bios[bios['name'].str.startswith("Keith")]
keith_first_names.head()

athlete_id           name   born_date  born_city       born_region  \
1897         1907   Keith Hanlon  1966-09-01        NaN               NaN   
3505         3517  Keith Wallace  1961-03-29    Preston           England   
6228         6255  Keith Hartley  1940-10-15  Vancouver  British Columbia   
8898         8946    Keith Mwila  1966-01-01        NaN               NaN   
12053       12118   Keith Hervey  1898-11-03     Fulham           England   

      born_country            NOC  height_cm  weight_kg   died_date  
1897           NaN        Ireland        NaN        NaN         NaN  
3505           GBR  Great Britain      165.0       51.0  1999-12-31  
6228           CAN         Canada      200.0       85.0         NaN  
8898           NaN         Zambia        NaN        NaN  1993-01-09  
12053          GBR  Great Britain        NaN        NaN  1973-02-22

In [29]:
#you can also use query function for filtereing
usa_player = bios.query('name =="Keith" and born_country == "USA"')
usa_player.head()

Empty DataFrame
Columns: [athlete_id, name, born_date, born_city, born_region, born_country, NOC, height_cm, weight_kg, died_date]
Index: []

In [30]:
#adding new columns
coffee["price"] = 4.99
coffee.head()

Day Coffee Type  Units Sold  price
0     Monday    Espresso          25   4.99
1     Monday       Latte          15   4.99
2    Tuesday    Espresso         100   4.99
3    Tuesday    Espresso         100   4.99
4  Wednesday    Espresso         150   4.99

In [31]:
import numpy as np
coffee['new_price'] = np.where(coffee['Coffee Type'] =='Espresso', 3.99, 5.99)
coffee.head()

Day Coffee Type  Units Sold  price  new_price
0     Monday    Espresso          25   4.99       3.99
1     Monday       Latte          15   4.99       5.99
2    Tuesday    Espresso         100   4.99       3.99
3    Tuesday    Espresso         100   4.99       3.99
4  Wednesday    Espresso         150   4.99       3.99

In [32]:
coffee.drop('price', axis =1) # but this is not inplace

Day Coffee Type  Units Sold  new_price
0      Monday    Espresso          25       3.99
1      Monday       Latte          15       5.99
2     Tuesday    Espresso         100       3.99
3     Tuesday    Espresso         100       3.99
4   Wednesday    Espresso         150       3.99
5   Wednesday       Latte          25       5.99
6    Thursday    Espresso          40       3.99
7    Thursday       Latte          30       5.99
8      Friday    Espresso          45       3.99
9      Friday       Latte          35       5.99
10   Saturday    Espresso          45       3.99
11   Saturday       Latte          35       5.99
12     Sunday    Espresso          45       3.99
13     Sunday       Latte          35       5.99

In [33]:
coffee.head()

Day Coffee Type  Units Sold  price  new_price
0     Monday    Espresso          25   4.99       3.99
1     Monday       Latte          15   4.99       5.99
2    Tuesday    Espresso         100   4.99       3.99
3    Tuesday    Espresso         100   4.99       3.99
4  Wednesday    Espresso         150   4.99       3.99

In [34]:
coffee.drop('price', axis = 1 , inplace = True)
coffee.head()

Day Coffee Type  Units Sold  new_price
0     Monday    Espresso          25       3.99
1     Monday       Latte          15       5.99
2    Tuesday    Espresso         100       3.99
3    Tuesday    Espresso         100       3.99
4  Wednesday    Espresso         150       3.99

In [35]:
#pandas way of creating a copy of dataframe
coffee_copy = coffee 
# if you do something like this - any change that you make on coffee_copy will get reflected to coffee as well
# Also any change that you make on coffee will get reflected to coffee_copy as well
#because both are just references to the same dataframe in memory
coffee_copy = coffee.copy()  # this way you create a true copy of the dataframe

In [36]:
coffee_copy['price'] = 6.99
coffee_copy.head()

Day Coffee Type  Units Sold  new_price  price
0     Monday    Espresso          25       3.99   6.99
1     Monday       Latte          15       5.99   6.99
2    Tuesday    Espresso         100       3.99   6.99
3    Tuesday    Espresso         100       3.99   6.99
4  Wednesday    Espresso         150       3.99   6.99

In [37]:
coffee.head()

Day Coffee Type  Units Sold  new_price
0     Monday    Espresso          25       3.99
1     Monday       Latte          15       5.99
2    Tuesday    Espresso         100       3.99
3    Tuesday    Espresso         100       3.99
4  Wednesday    Espresso         150       3.99

In [38]:
coffee['revenue'] = coffee['Units Sold'] * coffee['new_price']
coffee.head()

Day Coffee Type  Units Sold  new_price  revenue
0     Monday    Espresso          25       3.99    99.75
1     Monday       Latte          15       5.99    89.85
2    Tuesday    Espresso         100       3.99   399.00
3    Tuesday    Espresso         100       3.99   399.00
4  Wednesday    Espresso         150       3.99   598.50

In [39]:
#renaming columns
coffee.rename(columns = {'new_price':'price'}, inplace = True)
coffee.head()

Day Coffee Type  Units Sold  price  revenue
0     Monday    Espresso          25   3.99    99.75
1     Monday       Latte          15   5.99    89.85
2    Tuesday    Espresso         100   3.99   399.00
3    Tuesday    Espresso         100   3.99   399.00
4  Wednesday    Espresso         150   3.99   598.50

In [40]:
bios_new = bios.copy()
bios_new['first_name'] = bios_new['name'].str.split().str[0]
bios_new.head()

athlete_id                   name   born_date    born_city  \
0           1  Jean-François Blanchy  1886-12-12     Bordeaux   
1           2         Arnaud Boetsch  1969-04-01       Meulan   
2           3           Jean Borotra  1898-08-13     Biarritz   
3           4        Jacques Brugnon  1895-05-11  Paris VIIIe   
4           5           Albert Canet  1878-04-17   Wandsworth   

            born_region born_country     NOC  height_cm  weight_kg  \
0               Gironde          FRA  France        NaN        NaN   
1              Yvelines          FRA  France      183.0       76.0   
2  Pyrénées-Atlantiques          FRA  France      183.0       76.0   
3                 Paris          FRA  France      168.0       64.0   
4               England          GBR  France        NaN        NaN   

    died_date     first_name  
0  1960-10-02  Jean-François  
1         NaN         Arnaud  
2  1994-07-17           Jean  
3  1978-03-20        Jacques  
4  1930-07-25         Albert

In [41]:
bios_new['born_datetime'] = pd.to_datetime(bios_new['born_date'])

In [42]:
bios_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145500 entries, 0 to 145499
Data columns (total 12 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   athlete_id     145500 non-null  int64         
 1   name           145500 non-null  object        
 2   born_date      143693 non-null  object        
 3   born_city      110908 non-null  object        
 4   born_region    110908 non-null  object        
 5   born_country   110908 non-null  object        
 6   NOC            145499 non-null  object        
 7   height_cm      106651 non-null  float64       
 8   weight_kg      102070 non-null  float64       
 9   died_date      33940 non-null   object        
 10  first_name     145500 non-null  object        
 11  born_datetime  143693 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(8)
memory usage: 13.3+ MB


In [43]:
bios_new['born_year'] = bios_new['born_datetime'].dt.year
bios_new[['first_name','born_year']]

first_name  born_year
0       Jean-François     1886.0
1              Arnaud     1969.0
2                Jean     1898.0
3             Jacques     1895.0
4              Albert     1878.0
...               ...        ...
145495         Polina     2002.0
145496       Valeriya     1999.0
145497         Yuliya     1998.0
145498          André     1899.0
145499           Bill     1913.0

[145500 rows x 2 columns]

In [44]:
#save to csv
bios_new.to_csv('data/bios_new.csv', index = False)

In [45]:
#lamda functions on columns
bios['height_category'] = bios['height_cm'].apply(lambda x : 'Tall' if x > 200 else ('Average' if x > 175 else 'Short'))
bios[['name','height_cm','height_category']]

name  height_cm height_category
0       Jean-François Blanchy        NaN           Short
1              Arnaud Boetsch      183.0         Average
2                Jean Borotra      183.0         Average
3             Jacques Brugnon      168.0           Short
4                Albert Canet        NaN           Short
...                       ...        ...             ...
145495      Polina Luchnikova      167.0           Short
145496    Valeriya Merkusheva      168.0           Short
145497        Yuliya Smirnova      163.0           Short
145498         André Foussard      166.0           Short
145499          Bill Phillips        NaN           Short

[145500 rows x 3 columns]

In [46]:
# apply using custom functions
def athlete_category(row ):
    if row['height_cm'] > 200:
        return 'Tall'
    elif row['height_cm'] > 175:
        return 'Average'
    else:
        return 'Short'

In [47]:
bios['Category_height'] = bios.apply(athlete_category, axis =1)
bios[['name','height_cm','Category_height']]

name  height_cm Category_height
0       Jean-François Blanchy        NaN           Short
1              Arnaud Boetsch      183.0         Average
2                Jean Borotra      183.0         Average
3             Jacques Brugnon      168.0           Short
4                Albert Canet        NaN           Short
...                       ...        ...             ...
145495      Polina Luchnikova      167.0           Short
145496    Valeriya Merkusheva      168.0           Short
145497        Yuliya Smirnova      163.0           Short
145498         André Foussard      166.0           Short
145499          Bill Phillips        NaN           Short

[145500 rows x 3 columns]

In [48]:
nocs = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/complete-pandas-tutorial/refs/heads/master/data/noc_regions.csv')
nocs.head()

NOC       region                 notes
0  AFG  Afghanistan                   NaN
1  AHO      Curacao  Netherlands Antilles
2  ALB      Albania                   NaN
3  ALG      Algeria                   NaN
4  AND      Andorra                   NaN

In [49]:
#merging dataframes
bios = pd.merge(bios,nocs, left_on = 'born_country', right_on = 'NOC', how = 'left')
bios.head()

athlete_id                   name   born_date    born_city  \
0           1  Jean-François Blanchy  1886-12-12     Bordeaux   
1           2         Arnaud Boetsch  1969-04-01       Meulan   
2           3           Jean Borotra  1898-08-13     Biarritz   
3           4        Jacques Brugnon  1895-05-11  Paris VIIIe   
4           5           Albert Canet  1878-04-17   Wandsworth   

            born_region born_country   NOC_x  height_cm  weight_kg  \
0               Gironde          FRA  France        NaN        NaN   
1              Yvelines          FRA  France      183.0       76.0   
2  Pyrénées-Atlantiques          FRA  France      183.0       76.0   
3                 Paris          FRA  France      168.0       64.0   
4               England          GBR  France        NaN        NaN   

    died_date height_category Category_height NOC_y  region notes  
0  1960-10-02           Short           Short   FRA  France   NaN  
1         NaN         Average         Average   FRA  France   NaN  
2  1994-07-17         Average         Average   FRA  France   NaN  
3  1978-03-20           Short           Short   FRA  France   NaN  
4  1930-07-25           Short           Short   GBR      UK   NaN

In [50]:
#bios =bios.rename(columns = {'region': 'born_country_full'}, inplace = True) -- This is wrong
# when you do inplace = True, the function returns None. So you should not assign it back to bios
bios.rename(columns = {'region': 'born_country_full'}, inplace = True)
bios.head()

athlete_id                   name   born_date    born_city  \
0           1  Jean-François Blanchy  1886-12-12     Bordeaux   
1           2         Arnaud Boetsch  1969-04-01       Meulan   
2           3           Jean Borotra  1898-08-13     Biarritz   
3           4        Jacques Brugnon  1895-05-11  Paris VIIIe   
4           5           Albert Canet  1878-04-17   Wandsworth   

            born_region born_country   NOC_x  height_cm  weight_kg  \
0               Gironde          FRA  France        NaN        NaN   
1              Yvelines          FRA  France      183.0       76.0   
2  Pyrénées-Atlantiques          FRA  France      183.0       76.0   
3                 Paris          FRA  France      168.0       64.0   
4               England          GBR  France        NaN        NaN   

    died_date height_category Category_height NOC_y born_country_full notes  
0  1960-10-02           Short           Short   FRA            France   NaN  
1         NaN         Average         Average   FRA            France   NaN  
2  1994-07-17         Average         Average   FRA            France   NaN  
3  1978-03-20           Short           Short   FRA            France   NaN  
4  1930-07-25           Short           Short   GBR                UK   NaN

In [51]:
#You can see NOC_x and NOC_y columns after the merge. The suffix x and y are added to differentiate between the two NOC columns from the two dataframes. 
#You cn change the suffixes using the suffixes parameter in the merge function.
#bios = pd.merge(bios,nocs, left_on = 'citizenship', right_on = 'NOC', how = 'left', suffixes=('_bios','_noc'))

In [52]:
#concatenation of dataframes
usa_players = bios[bios["born_country"] == "USA"]
gbr_players = bios[bios["born_country"] == "GBR"]

usa_gbr_players = pd.concat([usa_players, gbr_players])
usa_gbr_players

athlete_id                name   born_date         born_city  \
54              55       Monique Javer  1967-07-22        Burlingame   
960            964    Xóchitl Escobedo  1968-09-17       West Covina   
961            965   Angélica Gavaldón  1973-10-03         El Centro   
1231          1238      Bert Schneider  1897-07-01         Cleveland   
1345          1352          Laura Berg  1975-01-06  Santa Fe Springs   
...            ...                 ...         ...               ...   
144811      148512  Benjamin Alexander  1983-05-08            London   
144815      148517       Ashley Watson  1993-10-28      Peterborough   
145005      148716     Peder Kongshaug  2001-08-13         Wimbledon   
145319      149041          Axel Brown  1992-04-02         Harrogate   
145388      149111      Jean-Luc Baker  1993-10-07           Burnley   

       born_region born_country                NOC_x  height_cm  weight_kg  \
54      California          USA        Great Britain      177.0       64.0   
960     California          USA               Mexico      170.0       60.0   
961     California          USA               Mexico      160.0       54.0   
1231          Ohio          USA               Canada        NaN        NaN   
1345    California          USA        United States      168.0       61.0   
...            ...          ...                  ...        ...        ...   
144811     England          GBR              Jamaica        NaN        NaN   
144815     England          GBR              Jamaica        NaN        NaN   
145005     England          GBR               Norway      184.0       86.0   
145319     England          GBR  Trinidad and Tobago        NaN        NaN   
145388     England          GBR        United States        NaN        NaN   

         died_date height_category Category_height NOC_y born_country_full  \
54             NaN         Average         Average   USA               USA   
960            NaN           Short           Short   USA               USA   
961            NaN           Short           Short   USA               USA   
1231    1986-02-20           Short           Short   USA               USA   
1345           NaN           Short           Short   USA               USA   
...            ...             ...             ...   ...               ...   
144811         NaN           Short           Short   GBR                UK   
144815         NaN           Short           Short   GBR                UK   
145005         NaN         Average         Average   GBR                UK   
145319         NaN           Short           Short   GBR                UK   
145388         NaN           Short           Short   GBR                UK   

       notes  
54       NaN  
960      NaN  
961      NaN  
1231     NaN  
1345     NaN  
...      ...  
144811   NaN  
144815   NaN  
145005   NaN  
145319   NaN  
145388   NaN  

[15433 rows x 15 columns]

Handling Null Values

In [53]:
coffee.head()

Day Coffee Type  Units Sold  price  revenue
0     Monday    Espresso          25   3.99    99.75
1     Monday       Latte          15   5.99    89.85
2    Tuesday    Espresso         100   3.99   399.00
3    Tuesday    Espresso         100   3.99   399.00
4  Wednesday    Espresso         150   3.99   598.50

In [54]:
coffee.loc[0:2, 'Units Sold'] = np.nan
coffee.head()

Day Coffee Type  Units Sold  price  revenue
0     Monday    Espresso         NaN   3.99    99.75
1     Monday       Latte         NaN   5.99    89.85
2    Tuesday    Espresso         NaN   3.99   399.00
3    Tuesday    Espresso       100.0   3.99   399.00
4  Wednesday    Espresso       150.0   3.99   598.50

In [55]:
coffee.info()
#units sold has 11 non null whereas others have 14

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Day          14 non-null     object 
 1   Coffee Type  14 non-null     object 
 2   Units Sold   11 non-null     float64
 3   price        14 non-null     float64
 4   revenue      14 non-null     float64
dtypes: float64(3), object(2)
memory usage: 692.0+ bytes


In [56]:
coffee.isna() #goes through each cell of the df and evaluates if it is null or not

Day  Coffee Type  Units Sold  price  revenue
0   False        False        True  False    False
1   False        False        True  False    False
2   False        False        True  False    False
3   False        False       False  False    False
4   False        False       False  False    False
5   False        False       False  False    False
6   False        False       False  False    False
7   False        False       False  False    False
8   False        False       False  False    False
9   False        False       False  False    False
10  False        False       False  False    False
11  False        False       False  False    False
12  False        False       False  False    False
13  False        False       False  False    False

In [57]:
coffee.isna().sum()
#column wise number of nulls

Day            0
Coffee Type    0
Units Sold     3
price          0
revenue        0
dtype: int64

In [58]:
#how to fill up nan 
coffee.fillna(100)

Day Coffee Type  Units Sold  price  revenue
0      Monday    Espresso       100.0   3.99    99.75
1      Monday       Latte       100.0   5.99    89.85
2     Tuesday    Espresso       100.0   3.99   399.00
3     Tuesday    Espresso       100.0   3.99   399.00
4   Wednesday    Espresso       150.0   3.99   598.50
5   Wednesday       Latte        25.0   5.99   149.75
6    Thursday    Espresso        40.0   3.99   159.60
7    Thursday       Latte        30.0   5.99   179.70
8      Friday    Espresso        45.0   3.99   179.55
9      Friday       Latte        35.0   5.99   209.65
10   Saturday    Espresso        45.0   3.99   179.55
11   Saturday       Latte        35.0   5.99   209.65
12     Sunday    Espresso        45.0   3.99   179.55
13     Sunday       Latte        35.0   5.99   209.65

In [59]:
coffee.fillna(coffee['Units Sold'].mean())

Day Coffee Type  Units Sold  price  revenue
0      Monday    Espresso   53.181818   3.99    99.75
1      Monday       Latte   53.181818   5.99    89.85
2     Tuesday    Espresso   53.181818   3.99   399.00
3     Tuesday    Espresso  100.000000   3.99   399.00
4   Wednesday    Espresso  150.000000   3.99   598.50
5   Wednesday       Latte   25.000000   5.99   149.75
6    Thursday    Espresso   40.000000   3.99   159.60
7    Thursday       Latte   30.000000   5.99   179.70
8      Friday    Espresso   45.000000   3.99   179.55
9      Friday       Latte   35.000000   5.99   209.65
10   Saturday    Espresso   45.000000   3.99   179.55
11   Saturday       Latte   35.000000   5.99   209.65
12     Sunday    Espresso   45.000000   3.99   179.55
13     Sunday       Latte   35.000000   5.99   209.65

In [60]:
#we can use something like - coffee['Units Sold'].interpolate() - to see the pattern of how units sold is progressing on a day to day basis
# And accordingly fill up the value
# However this only works when we are working with null values in the middle of the dataset and not in the beginning
# so in ourcase it might not work -but let's still try it 

coffee['Units Sold'].interpolate() # It didn't work

0       NaN
1       NaN
2       NaN
3     100.0
4     150.0
5      25.0
6      40.0
7      30.0
8      45.0
9      35.0
10     45.0
11     35.0
12     45.0
13     35.0
Name: Units Sold, dtype: float64

In [61]:
#we can just drop the rows with missing value
coffee.dropna()
coffee # if you do this you won't see any changes because by default it is not inplace
coffee.dropna(inplace=True)
coffee

Day Coffee Type  Units Sold  price  revenue
3     Tuesday    Espresso       100.0   3.99   399.00
4   Wednesday    Espresso       150.0   3.99   598.50
5   Wednesday       Latte        25.0   5.99   149.75
6    Thursday    Espresso        40.0   3.99   159.60
7    Thursday       Latte        30.0   5.99   179.70
8      Friday    Espresso        45.0   3.99   179.55
9      Friday       Latte        35.0   5.99   209.65
10   Saturday    Espresso        45.0   3.99   179.55
11   Saturday       Latte        35.0   5.99   209.65
12     Sunday    Espresso        45.0   3.99   179.55
13     Sunday       Latte        35.0   5.99   209.65

#Aggregating the data

In [62]:
bios.head()

athlete_id                   name   born_date    born_city  \
0           1  Jean-François Blanchy  1886-12-12     Bordeaux   
1           2         Arnaud Boetsch  1969-04-01       Meulan   
2           3           Jean Borotra  1898-08-13     Biarritz   
3           4        Jacques Brugnon  1895-05-11  Paris VIIIe   
4           5           Albert Canet  1878-04-17   Wandsworth   

            born_region born_country   NOC_x  height_cm  weight_kg  \
0               Gironde          FRA  France        NaN        NaN   
1              Yvelines          FRA  France      183.0       76.0   
2  Pyrénées-Atlantiques          FRA  France      183.0       76.0   
3                 Paris          FRA  France      168.0       64.0   
4               England          GBR  France        NaN        NaN   

    died_date height_category Category_height NOC_y born_country_full notes  
0  1960-10-02           Short           Short   FRA            France   NaN  
1         NaN         Average         Average   FRA            France   NaN  
2  1994-07-17         Average         Average   FRA            France   NaN  
3  1978-03-20           Short           Short   FRA            France   NaN  
4  1930-07-25           Short           Short   GBR                UK   NaN

In [63]:
bios['born_city'].value_counts()

born_city
Budapest           1378
Moskva (Moscow)     883
Oslo                708
Stockholm           629
Praha (Prague)      600
                   ... 
Bodrogkisfalud        1
Ternberg              1
Klaus                 1
Plaški                1
Dulwich Hill          1
Name: count, Length: 22368, dtype: int64

In [64]:
bios[bios['born_country'] == "USA"]['born_region'].value_counts()

born_region
California              1634
New York                 990
Illinois                 585
Pennsylvania             530
Massachusetts            530
New Jersey               381
Texas                    368
Minnesota                365
Ohio                     328
Michigan                 319
Washington               240
Florida                  235
Wisconsin                209
Colorado                 207
Connecticut              156
Indiana                  150
Oregon                   132
Georgia                  129
Virginia                 121
Maryland                 117
District of Columbia     107
Iowa                     102
Hawaiʻi                   95
Kansas                    94
Oklahoma                  93
Louisiana                 92
Utah                      91
Missouri                  91
North Carolina            86
Arizona                   83
New Hampshire             83
Vermont                   68
Mississippi               66
Alabama                   64
Ke

In [65]:
coffee

Day Coffee Type  Units Sold  price  revenue
3     Tuesday    Espresso       100.0   3.99   399.00
4   Wednesday    Espresso       150.0   3.99   598.50
5   Wednesday       Latte        25.0   5.99   149.75
6    Thursday    Espresso        40.0   3.99   159.60
7    Thursday       Latte        30.0   5.99   179.70
8      Friday    Espresso        45.0   3.99   179.55
9      Friday       Latte        35.0   5.99   209.65
10   Saturday    Espresso        45.0   3.99   179.55
11   Saturday       Latte        35.0   5.99   209.65
12     Sunday    Espresso        45.0   3.99   179.55
13     Sunday       Latte        35.0   5.99   209.65

In [66]:
coffee.groupby('Coffee Type')
#this will just create a group by object but not print anything
#Because groupby() is lazy.

#It waits for an aggregation, transformation, or filtering operation like mean(), sum() etc

coffee.groupby('Coffee Type').sum()


Day  Units Sold  price  \
Coffee Type                                                                    
Espresso     TuesdayWednesdayThursdayFridaySaturdaySunday       425.0  23.94   
Latte               WednesdayThursdayFridaySaturdaySunday       160.0  29.95   

             revenue  
Coffee Type           
Espresso     1695.75  
Latte         958.40

In [67]:
#How to display different aggregates based on the columns
coffee.groupby('Coffee Type').agg({'Units Sold': 'sum', 'price':'mean'})

Units Sold  price
Coffee Type                   
Espresso          425.0   3.99
Latte             160.0   5.99

In [68]:
#we can also groupby multiple attributes
coffee.groupby(['Coffee Type','Day']).agg({'Units Sold': 'sum', 'price':'mean'})
#not that useful in our case cuz we have only unique days

Units Sold  price
Coffee Type Day                         
Espresso    Friday           45.0   3.99
            Saturday         45.0   3.99
            Sunday           45.0   3.99
            Thursday         40.0   3.99
            Tuesday         100.0   3.99
            Wednesday       150.0   3.99
Latte       Friday           35.0   5.99
            Saturday         35.0   5.99
            Sunday           35.0   5.99
            Thursday         30.0   5.99
            Wednesday        25.0   5.99

In [69]:
coffee

Day Coffee Type  Units Sold  price  revenue
3     Tuesday    Espresso       100.0   3.99   399.00
4   Wednesday    Espresso       150.0   3.99   598.50
5   Wednesday       Latte        25.0   5.99   149.75
6    Thursday    Espresso        40.0   3.99   159.60
7    Thursday       Latte        30.0   5.99   179.70
8      Friday    Espresso        45.0   3.99   179.55
9      Friday       Latte        35.0   5.99   209.65
10   Saturday    Espresso        45.0   3.99   179.55
11   Saturday       Latte        35.0   5.99   209.65
12     Sunday    Espresso        45.0   3.99   179.55
13     Sunday       Latte        35.0   5.99   209.65

In [70]:
#working with pivot table
pivot_coffee = coffee.pivot(columns='Coffee Type', index= 'Day',values= 'revenue' )
pivot_coffee

Coffee Type  Espresso   Latte
Day                          
Friday         179.55  209.65
Saturday       179.55  209.65
Sunday         179.55  209.65
Thursday       159.60  179.70
Tuesday        399.00     NaN
Wednesday      598.50  149.75

In [71]:
bios['born_date'] = pd.to_datetime(bios['born_date'])
bios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145500 entries, 0 to 145499
Data columns (total 15 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   athlete_id         145500 non-null  int64         
 1   name               145500 non-null  object        
 2   born_date          143693 non-null  datetime64[ns]
 3   born_city          110908 non-null  object        
 4   born_region        110908 non-null  object        
 5   born_country       110908 non-null  object        
 6   NOC_x              145499 non-null  object        
 7   height_cm          106651 non-null  float64       
 8   weight_kg          102070 non-null  float64       
 9   died_date          33940 non-null   object        
 10  height_category    145500 non-null  object        
 11  Category_height    145500 non-null  object        
 12  NOC_y              110639 non-null  object        
 13  born_country_full  110636 non-null  object  

In [72]:
#lets see number of player based on age
bios.groupby(bios['born_date'].dt.year)['name'].count().reset_index().sort_values(by = 'name',ascending = False)

born_date  name
139     1972.0  2231
152     1985.0  2227
140     1973.0  2216
138     1971.0  2205
137     1970.0  2174
..         ...   ...
5       1838.0     1
4       1837.0     1
3       1836.0     1
2       1833.0     1
176     2009.0     1

[177 rows x 2 columns]

Advanced functionaliies

In [74]:
#shift # rolling # rank

coffee['yesterday_revenue'] = coffee['revenue'].shift(2)
coffee

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue
3     Tuesday    Espresso       100.0   3.99   399.00                NaN
4   Wednesday    Espresso       150.0   3.99   598.50                NaN
5   Wednesday       Latte        25.0   5.99   149.75             399.00
6    Thursday    Espresso        40.0   3.99   159.60             598.50
7    Thursday       Latte        30.0   5.99   179.70             149.75
8      Friday    Espresso        45.0   3.99   179.55             159.60
9      Friday       Latte        35.0   5.99   209.65             179.70
10   Saturday    Espresso        45.0   3.99   179.55             179.55
11   Saturday       Latte        35.0   5.99   209.65             209.65
12     Sunday    Espresso        45.0   3.99   179.55             179.55
13     Sunday       Latte        35.0   5.99   209.65             209.65

In [75]:
coffee.loc[5,'yesterday_revenue'] = np.nan

In [76]:
coffee

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue
3     Tuesday    Espresso       100.0   3.99   399.00                NaN
4   Wednesday    Espresso       150.0   3.99   598.50                NaN
5   Wednesday       Latte        25.0   5.99   149.75                NaN
6    Thursday    Espresso        40.0   3.99   159.60             598.50
7    Thursday       Latte        30.0   5.99   179.70             149.75
8      Friday    Espresso        45.0   3.99   179.55             159.60
9      Friday       Latte        35.0   5.99   209.65             179.70
10   Saturday    Espresso        45.0   3.99   179.55             179.55
11   Saturday       Latte        35.0   5.99   209.65             209.65
12     Sunday    Espresso        45.0   3.99   179.55             179.55
13     Sunday       Latte        35.0   5.99   209.65             209.65

In [78]:
coffee['pct_change'] = coffee['revenue']/coffee['yesterday_revenue'] *100
coffee

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue  \
3     Tuesday    Espresso       100.0   3.99   399.00                NaN   
4   Wednesday    Espresso       150.0   3.99   598.50                NaN   
5   Wednesday       Latte        25.0   5.99   149.75                NaN   
6    Thursday    Espresso        40.0   3.99   159.60             598.50   
7    Thursday       Latte        30.0   5.99   179.70             149.75   
8      Friday    Espresso        45.0   3.99   179.55             159.60   
9      Friday       Latte        35.0   5.99   209.65             179.70   
10   Saturday    Espresso        45.0   3.99   179.55             179.55   
11   Saturday       Latte        35.0   5.99   209.65             209.65   
12     Sunday    Espresso        45.0   3.99   179.55             179.55   
13     Sunday       Latte        35.0   5.99   209.65             209.65   

    pct_change  
3          NaN  
4          NaN  
5          NaN  
6    26.666667  
7   120.000000  
8   112.500000  
9   116.666667  
10  100.000000  
11  100.000000  
12  100.000000  
13  100.000000

In [79]:
#we can also shift backwards by using shift(-2)

In [80]:
#rank
bios

athlete_id                   name  born_date        born_city  \
0                1  Jean-François Blanchy 1886-12-12         Bordeaux   
1                2         Arnaud Boetsch 1969-04-01           Meulan   
2                3           Jean Borotra 1898-08-13         Biarritz   
3                4        Jacques Brugnon 1895-05-11      Paris VIIIe   
4                5           Albert Canet 1878-04-17       Wandsworth   
...            ...                    ...        ...              ...   
145495      149222      Polina Luchnikova 2002-01-30            Serov   
145496      149223    Valeriya Merkusheva 1999-09-20  Moskva (Moscow)   
145497      149224        Yuliya Smirnova 1998-05-08           Kotlas   
145498      149225         André Foussard 1899-05-19            Niort   
145499      149814          Bill Phillips 1913-07-15     Dulwich Hill   

                 born_region born_country      NOC_x  height_cm  weight_kg  \
0                    Gironde          FRA     France        NaN        NaN   
1                   Yvelines          FRA     France      183.0       76.0   
2       Pyrénées-Atlantiques          FRA     France      183.0       76.0   
3                      Paris          FRA     France      168.0       64.0   
4                    England          GBR     France        NaN        NaN   
...                      ...          ...        ...        ...        ...   
145495            Sverdlovsk          RUS        ROC      167.0       61.0   
145496                Moskva          RUS        ROC      168.0       65.0   
145497           Arkhangelsk          RUS        ROC      163.0       55.0   
145498           Deux-Sèvres          FRA     France      166.0        NaN   
145499       New South Wales          AUS  Australia        NaN        NaN   

         died_date height_category Category_height NOC_y born_country_full  \
0       1960-10-02           Short           Short   FRA            France   
1              NaN         Average         Average   FRA            France   
2       1994-07-17         Average         Average   FRA            France   
3       1978-03-20           Short           Short   FRA            France   
4       1930-07-25           Short           Short   GBR                UK   
...            ...             ...             ...   ...               ...   
145495         NaN           Short           Short   RUS            Russia   
145496         NaN           Short           Short   RUS            Russia   
145497         NaN           Short           Short   RUS            Russia   
145498  1986-03-18           Short           Short   FRA            France   
145499  2003-10-20           Short           Short   AUS         Australia   

       notes  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  
...      ...  
145495   NaN  
145496   NaN  
145497   NaN  
145498   NaN  
145499   NaN  

[145500 rows x 15 columns]

In [82]:
bios['height_rank'] = bios['height_cm'].rank()
bios[['name', 'height_cm', 'height_rank']].sort_values(by = 'height_rank', ascending=False)

name  height_cm  height_rank
89070                Yao Ming      226.0     106651.0
5781           Tommy Burleson      223.0     106649.5
6978          Arvydas Sabonis      223.0     106649.5
89075          Roberto Dueñas      221.0     106647.0
120266           Zhang Zhaoxu      221.0     106647.0
...                       ...        ...          ...
145490            Sin Ye-Chan        NaN          NaN
145491          Matthew Wepke        NaN          NaN
145492  Carlos García-Ordóñez        NaN          NaN
145493     Landysh Falyakhova        NaN          NaN
145499          Bill Phillips        NaN          NaN

[145500 rows x 3 columns]

Good catch — this is **expected pandas behavior** 👍
Your `height_rank` is not an integer because of **ties** and the **default ranking method**.

---

## What’s happening

You used:

```python
bios['height_rank'] = bios['height_cm'].rank()
```

By default, `Series.rank()` uses:

```text
method='average'
```

### Meaning of `method='average'`

When **multiple rows have the same value**, pandas assigns them the **average of the ranks they would occupy**.

#### Example from your output

```text
Tommy Burleson   223 cm → rank 106649.5
Arvydas Sabonis  223 cm → rank 106649.5
```

Why `106649.5`?

* Suppose these two heights should occupy ranks **106649 and 106650**
* Pandas assigns:
  [
  (106649 + 106650) / 2 = 106649.5
  ]

That’s why you see `.5`.

---

## Why Yao Ming has an integer rank

```text
Yao Ming 226 cm → 106651.0
```

He has a **unique height**, so no averaging is needed.

---

## How to get integer ranks (your options)

### ✅ Option 1: Dense ranking (most common)

No gaps, integers only.

```python
bios['height_rank'] = bios['height_cm'].rank(method='dense', ascending=False)
```

Example:

```text
226 → 1
223 → 2
221 → 3
```

---

### ✅ Option 2: Competition ranking (sports-style)

Keeps gaps.

```python
bios['height_rank'] = bios['height_cm'].rank(method='min', ascending=False)
```

Example:

```text
226 → 1
223 → 2
223 → 2
221 → 4
```

---

### ✅ Option 3: Order-based (no ties)

Ranks by row order.

```python
bios['height_rank'] = bios['height_cm'].rank(method='first', ascending=False)
```

---

### ✅ Option 4: Explicit integer cast (⚠️ not recommended alone)

```python
bios['height_rank'] = bios['height_cm'].rank().astype(int)
```

This **silently truncates** and can be misleading.

---

## Handling NaNs explicitly (optional)

```python
bios['height_rank'] = bios['height_cm'].rank(
    method='dense',
    ascending=False,
    na_option='bottom'
)
```

---

## TL;DR

* `.5` happens because **ties are averaged by default**
* This is **correct and intentional**
* Use `method='dense'` if you want clean integer ranks

If you want, I can help you pick the **right ranking style** depending on whether this is for analytics, reporting, or modeling.


In [83]:
#rolling functions
coffee['cumulative_sum'] = coffee['revenue'].cumsum()
coffee

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue  \
3     Tuesday    Espresso       100.0   3.99   399.00                NaN   
4   Wednesday    Espresso       150.0   3.99   598.50                NaN   
5   Wednesday       Latte        25.0   5.99   149.75                NaN   
6    Thursday    Espresso        40.0   3.99   159.60             598.50   
7    Thursday       Latte        30.0   5.99   179.70             149.75   
8      Friday    Espresso        45.0   3.99   179.55             159.60   
9      Friday       Latte        35.0   5.99   209.65             179.70   
10   Saturday    Espresso        45.0   3.99   179.55             179.55   
11   Saturday       Latte        35.0   5.99   209.65             209.65   
12     Sunday    Espresso        45.0   3.99   179.55             179.55   
13     Sunday       Latte        35.0   5.99   209.65             209.65   

    pct_change  cumulative_sum  
3          NaN          399.00  
4          NaN          997.50  
5          NaN         1147.25  
6    26.666667         1306.85  
7   120.000000         1486.55  
8   112.500000         1666.10  
9   116.666667         1875.75  
10  100.000000         2055.30  
11  100.000000         2264.95  
12  100.000000         2444.50  
13  100.000000         2654.15

In [84]:
latte = coffee[coffee['Coffee Type']== "Latte"].copy()
latte

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue  \
5   Wednesday       Latte        25.0   5.99   149.75                NaN   
7    Thursday       Latte        30.0   5.99   179.70             149.75   
9      Friday       Latte        35.0   5.99   209.65             179.70   
11   Saturday       Latte        35.0   5.99   209.65             209.65   
13     Sunday       Latte        35.0   5.99   209.65             209.65   

    pct_change  cumulative_sum  
5          NaN         1147.25  
7   120.000000         1486.55  
9   116.666667         1875.75  
11  100.000000         2264.95  
13  100.000000         2654.15

In [85]:
latte['3day_selling_units'] = latte['Units Sold'].rolling(3).sum()
latte

Day Coffee Type  Units Sold  price  revenue  yesterday_revenue  \
5   Wednesday       Latte        25.0   5.99   149.75                NaN   
7    Thursday       Latte        30.0   5.99   179.70             149.75   
9      Friday       Latte        35.0   5.99   209.65             179.70   
11   Saturday       Latte        35.0   5.99   209.65             209.65   
13     Sunday       Latte        35.0   5.99   209.65             209.65   

    pct_change  cumulative_sum  3day_selling_units  
5          NaN         1147.25                 NaN  
7   120.000000         1486.55                 NaN  
9   116.666667         1875.75                90.0  
11  100.000000         2264.95               100.0  
13  100.000000         2654.15               105.0